driver/src/models.rs

```diff
@@ -21,9 +25,53 @@ pub trait Serializable {
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
 #[serde(rename_all = "camelCase")]
 pub struct State {
-  pub state_hash: String,
-  pub state_index: i32,
-  pub balances: Vec<u128>,
+    pub state_hash: String,
+    pub state_index: i32,
+    balances: Vec<u128>,
+}
+
+impl State {
+    pub fn new(state_hash: String, state_index: i32, balances: Vec<u128>) -> Self {
+        State { state_hash, state_index, balances }
+    }
+    fn balance_index(token_id: u8, account_id: u16) -> usize {
+        TOKENS as usize * account_id as usize  + token_id as usize
+    }
+    pub fn read_balance(&self, token_id: u8, account_id: u16) -> u128 {
+        self.balances[State::balance_index(token_id, account_id)]
+    }
+    pub fn increment_balance(&mut self, token_id: u8, account_id: u16, amount: u128) {
+        self.balances[State::balance_index(token_id, account_id)] += amount;
+    }
+    pub fn decrement_balance(&mut self, token_id: u8, account_id: u16, amount: u128) {
+        self.balances[State::balance_index(token_id, account_id)] -= amount;
+    }
+    pub fn serialize_balances(&self, num_tokens: u8) -> serde_json::Value {
+        assert_eq!(
+            self.balances.len() % num_tokens as usize, 0,
+            "Balance vector cannot be split into equal accounts"
+        );
+        let mut accounts: HashMap<String, HashMap<String, String>> = HashMap::new();
+        let mut current_account = 0;
+        for account_balances in self.balances.chunks(num_tokens as usize) {
+            accounts.insert(account_id(current_account), (0..num_tokens)
+                .map(token_id)
+                .zip(account_balances.iter().map(|b| b.to_string()))
+                .collect());
+            current_account += 1;
+        }
+        json!(accounts)
+    }
+
+    pub fn update_balances(&mut self, orders: &[Order], solution: &Solution) {
```

Comment:
I'm not entirely sure if this method should be part of the State trait. It makes it so that `State` has to know about `Solution`, which is a coupling that might not be necessary. E.g. for deposits and withdraws you apply them inside the driver (that is the `State` trait doesn't have any reference to the `PendingFlux` object).

Anyways, this is just architectural nitty gritty, if it becomes a problem in the future, we can easily put this method back into the driver.